pkg/admission/apibinding/apibinding_admission.go

```diff
@@ -0,0 +1,234 @@
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1/helper"
+)
+
+const (
+	PluginName = "apis.kcp.dev/APIBinding"
+)
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName,
+		func(_ io.Reader) (admission.Interface, error) {
+			return &apiBindingAdmission{
+				Handler:          admission.NewHandler(admission.Create, admission.Update),
+				createAuthorizer: kcpadmissionhelpers.NewAdmissionAuthorizer,
+			}, nil
+		})
+}
+
+type apiBindingAdmission struct {
+	*admission.Handler
+	kubeClusterClient *kubernetes.Cluster
+
+	createAuthorizer kcpadmissionhelpers.AdmissionAuthorizerFactory
+}
+
+// Ensure that the required admission interfaces are implemented.
+var _ = admission.ValidationInterface(&apiBindingAdmission{})
+var _ = admission.MutationInterface(&apiBindingAdmission{})
+var _ = admission.InitializationValidator(&apiBindingAdmission{})
+
+var phaseOrdinal = map[apisv1alpha1.APIBindingPhaseType]int{
+	apisv1alpha1.APIBindingPhaseType(""): 1,
+	apisv1alpha1.APIBindingPhaseBinding:  2,
+	apisv1alpha1.APIBindingPhaseBound:    3,
+	// not including Rebinding as you can move back and forth between Bound and Rebinding
+}
+
+// Validate validates the creation and updating of APIBinding resources. It also performs a SubjectAccessReview
+// making sure the user is allowed to use the 'bind' verb with the referenced APIExport.
+func (o *apiBindingAdmission) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) error {
+	if a.GetResource().GroupResource() != apisv1alpha1.Resource("apibindings") {
+		return nil
+	}
+
+	obj, err := kcpadmissionhelpers.NativeObject(a.GetObject())
+	if err != nil {
+		// nolint: nilerr
+		return nil // only work on unstructured APIBindings
+	}
+	apiBinding, ok := obj.(*apisv1alpha1.APIBinding)
+	if !ok {
+		// nolint: nilerr
+		return nil // only work on unstructured APIBindings
+	}
+
+	// Return early if there's nothing to validate
+	if apiBinding.Spec.Reference.Workspace == nil {
+		return nil
+	}
+
+	// Object validation
+	var errs field.ErrorList
+
+	switch a.GetOperation() {
+	case admission.Create:
+		errs = ValidateAPIBinding(apiBinding)
+	case admission.Update:
+		obj, err = kcpadmissionhelpers.NativeObject(a.GetOldObject())
+		if err != nil {
+			return fmt.Errorf("unexpected unknown old object, got %v, expected APIBinding", a.GetOldObject().GetObjectKind().GroupVersionKind().Kind)
+		}
+		old, ok := obj.(*apisv1alpha1.APIBinding)
+		if !ok {
+			return fmt.Errorf("unexpected unknown old object, got %v, expected APIBinding", obj.GetObjectKind().GroupVersionKind().Kind)
+		}
+		errs = ValidateAPIBindingUpdate(old, apiBinding)
+	}
+	if len(errs) > 0 {
+		return admission.NewForbidden(a, fmt.Errorf("%v", errs))
+	}
+
+	// Determine the cluster name for the referenced export
+	cluster, err := genericapirequest.ValidClusterFrom(ctx)
+	if err != nil {
+		// TODO(ncdc): is this user-facing error message ok?
```

Comment:
I'd think about it as - what did the user do to send this lcluster name? How could we structure the error to tell them what to do in order to fix it?

Reply:
@stevekuznetsov updated text, PTAL!